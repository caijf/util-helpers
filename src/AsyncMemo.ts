import { Cache, CacheOptions } from 'cache2';
import { isString, uniqueId } from 'ut2';

/**
 * 异步缓存
 *
 * @class
 * @see {@link https://www.npmjs.com/package/cache2 | cache2}
 * @param {Object} [options] 缓存配置项，更多配置项可参考 [`cache2`](https://www.npmjs.com/package/cache2)
 * @param {number} [options.max] 最大缓存数量
 * @param {'replaced' | 'limited'} [options.maxStrategy] 缓存策略
 */
class AsyncMemo<DataType = any> {
  private promiseCache: Record<string, Promise<any>>;
  /**
   * cache2 实例，用于管理缓存
   */
  cache: Cache<DataType>;

  constructor(options?: Partial<CacheOptions>) {
    this.promiseCache = {};
    this.cache = new Cache(uniqueId('uh_async_memo'), options);
  }

  /**
   * 执行异步方法
   *
   * @param {Function} asyncFn 异步方法
   * @param {string} [key] 缓存键，如果没有该值将直接执行异步方法。
   * @param {Object} [options] 配置项
   * @param {number} [options.ttl] 数据存活时间
   * @param {boolean} [options.persisted=true] 数据持久化，如果存在缓存数据，直接返回缓存数据，不再执行异步方法。<br/>即使不开启该配置，不影响在异步执行成功后缓存数据。
   * @returns {Promise<*>} 异步结果
   */
  run(
    asyncFn: (...args: any[]) => Promise<DataType>,
    key?: string,
    options?: {
      ttl?: number;
      persisted?: boolean;
    }
  ) {
    if (!key || !isString(key)) {
      return asyncFn();
    }

    const opts = {
      persisted: true,
      ...options
    };

    if (opts.persisted) {
      const data = this.cache.get(key);
      if (data) {
        return Promise.resolve(data);
      }
    }

    if (!this.promiseCache[key]) {
      this.promiseCache[key] = asyncFn()
        .then((res) => {
          delete this.promiseCache[key];
          this.cache.set(key, res, opts.ttl);
          return res;
        })
        .catch((err) => {
          delete this.promiseCache[key];
          return Promise.reject(err);
        });
    }
    return this.promiseCache[key] as Promise<DataType>;
  }
}

export default AsyncMemo;
