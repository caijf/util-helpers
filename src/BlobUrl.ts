import { findKey } from 'ut2';
import { createObjectURL, revokeObjectURL } from './utils/native';

/**
 * Blob 对象 URL 记录。
 *
 * 便于管理项目中上传文件通过 `URL.createObjectURL` 生成的对象URL。
 *
 * 特点：
 * 1. 避免重复创建，相同的 `Blob` 对象只会生成一个 URL 字符串
 * 2. 当清除缓存时，自动释放内存
 *
 * @class
 * @see {@link https://developer.mozilla.org/zh-CN/docs/Web/API/URL/createObjectURL_static | URL：createObjectURL()}
 * @example
 * const blobUrl = new BlobUrl();
 *
 */
class BlobUrl {
  protected cache: Map<Blob, string>;

  constructor() {
    this.cache = new Map<Blob, string>();
  }

  /**
   * 同 `URL.createObjectURL` 方法，创建一个对象 URL 字符串。
   *
   * 生成时内部会记录一个缓存，相同的 `Blob` 对象最多只会产生一个 URL 字符串。
   *
   * @param {Blob} obj 用于创建 URL 的 File、Blob 对象。
   * @returns 一个包含对象 URL 的字符串，可用于引用指定源 object 的内容。
   */
  createObjectURL(obj: Blob) {
    if (this.cache.has(obj)) {
      return this.cache.get(obj);
    }
    const url = createObjectURL(obj);
    if (url) {
      this.cache.set(obj, url);
    }
    return url;
  }

  /**
   * 同 `URL.revokeObjectURL` 方法，释放对象 URL 。
   *
   * @param {string} url 通过调用 `createObjectURL()` 方法创建的对象 URL 的字符串。
   */
  revokeObjectURL(url: string) {
    const obj = findKey(this.cache, (v) => v === url);
    if (obj) {
      this.cache.delete(obj);
    }
    revokeObjectURL(url);
  }

  /**
   * 清理缓存。
   *
   * 迭代调用 `URL.revokeObjectURL` 来释放内存。
   */
  clear() {
    this.cache.forEach((v) => {
      revokeObjectURL(v);
    });
    this.cache.clear();
  }
}

export default BlobUrl;
