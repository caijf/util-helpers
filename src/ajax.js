/**
 * @typedef {XMLHttpRequest['onloadstart']} XMLHttpRequestListener XMLHttpRequest 事件对象
 */

/**
 * @see {@link https://developer.mozilla.org/zh-CN/docs/Web/API/XMLHttpRequest|XMLHttpRequest}
 * @typedef {Object} AjaxOptions ajax配置项
 * @property {string} [method="get"] 创建请求时使用的方法
 * @property {boolean} [async=true] 是否异步执行操作
 * @property {string|null} [user=null] 用户名，用于认证用途
 * @property {string|null} [password=null] 密码，用于认证用途
 * @property {Object.<string, string>} [headers] 自定义请求头
 * @property {XMLHttpRequestResponseType} [responseType] 响应类型
 * @property {number} [timeout] 请求超时的毫秒数
 * @property {boolean} [withCredentials=false] 跨域请求时是否需要使用凭证
 * @property {Parameters<XMLHttpRequest['send']>[0]} [data=null] 请求体被发送的数据
 * @property {XMLHttpRequest['onreadystatechange']} [options.onReadyStateChange] 当 readyState 属性发生变化时触发
 * @property {XMLHttpRequestListener} [onLoadStart] 接收到响应数据时触发
 * @property {XMLHttpRequestListener} [onProgress] 请求接收到更多数据时，周期性地触发
 * @property {XMLHttpRequestListener} [onAbort] 当 request 被停止时触发，例如当程序调用 XMLHttpRequest.abort() 时
 * @property {XMLHttpRequestListener} [onTimeout] 在预设时间内没有接收到响应时触发
 * @property {XMLHttpRequestListener} [onError] 当 request 遭遇错误时触发
 * @property {XMLHttpRequestListener} [onLoad] 请求成功完成时触发
 * @property {XMLHttpRequestListener} [onLoadEnd] 请求结束时触发，无论请求成功 (load) 还是失败 (abort 或 error)
 */

/**
 * 请求<br/><br/>
 * 
 * <em style="font-weight: bold;">注意：该方法仅适用于浏览器端。</em>
 * 
 * @static
 * @alias module:Other.ajax
 * @since 4.16.0
 * @see {@link https://developer.mozilla.org/zh-CN/docs/Web/API/XMLHttpRequest|XMLHttpRequest}
 * @param {string} url 地址
 * @param {AjaxOptions} [options] 配置项
 * @returns {Promise<ProgressEvent>}
 * @example
 * ajax('/somefile').then(res=>{
 *   // do something
 * });
 * 
 * ajax('/api', { method: 'post' }).then(res=>{
 *   // do something
 * });
 */
function ajax(url, options) {
  const {
    method = 'get',
    data = null,
    timeout,
    headers,
    withCredentials = false,
    async = true,
    user = null,
    password = null,
    responseType,
    onReadyStateChange,
    onLoadStart,
    onProgress,
    onAbort,
    onTimeout,
    onError,
    onLoad,
    onLoadEnd
  } = options || {};

  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    xhr.open(method.toLowerCase(), url, async, user, password);

    // 设置请求超时
    if (typeof timeout === 'number' && timeout > 0) {
      xhr.timeout = timeout;
    }

    if (onReadyStateChange) {
      xhr.onreadystatechange = onReadyStateChange;
    }

    // 跨域请求时是否需要使用凭证
    xhr.withCredentials = withCredentials;

    // 设置响应类型
    if (responseType) {
      xhr.responseType = responseType;
    }

    // 设置请求头
    if (typeof headers === 'object') {
      Object.keys(headers).map(item => {
        xhr.setRequestHeader(item, headers[item]);
      });
    }

    /**
     * 请求成功异步调用
     * @param {XMLHttpRequestListener} [cb] 回调方法
     */
    const wrapSuccess = (cb) => {
      /**
       * 内部方法
       * @param {ProgressEvent} e 事件对象
       */
      return (e) => {
        resolve(e);
        cb?.call(xhr, e);
      }
    }

    /**
     * 请求失败（中断/超时/失败）处理
       * @param {XMLHttpRequestListener} [cb] 回调方法
     */
    const wrapError = (cb) => {
      /**
       * 内部方法
       * @param {ProgressEvent} e 事件对象
       */
      return (e) => {
        reject(e);
        cb?.call(xhr, e);
      }
    }

    // 事件处理
    /**@type {Object.<keyof XMLHttpRequestEventTargetEventMap, XMLHttpRequestListener | undefined>} */
    const events = {
      loadstart: onLoadStart,
      progress: onProgress,
      abort: wrapError(onAbort),
      timeout: wrapError(onTimeout),
      error: wrapError(onError),
      load: wrapSuccess(onLoad),
      loadend: onLoadEnd
    }
    /**@type {(keyof XMLHttpRequestEventTargetEventMap)[]} */
    // @ts-ignore
    const eventNames = Object.keys(events);

    eventNames.map(item => {
      const func = events[item];
      if (func) {
        xhr.addEventListener(item, func);
      }
    });

    xhr.send(data);
  });
}

export default ajax;