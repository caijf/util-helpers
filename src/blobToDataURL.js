//  file、blob文件如何预览图片？
//  方法1：将file或者blob类型文件转成base64数据，再作为src赋值给img标签
//  方法2：使用 window.URL.createObjectURL(blob) 为blob、file 创建一个指向该参数对象的URL

/**
 * 将 Blob 或 File 对象转成 data:URL 格式的 Base64 字符串<br/><br/>
 * 
 * <em style="font-weight: bold;">注意：该方法仅适用于浏览器端。</em>
 *
 * @static
 * @alias module:Processor.blobToDataURL
 * @see 参考 {@link https://developer.mozilla.org/zh-CN/docs/Web/API/FileReader/readAsDataURL|FileReader.readAsDataURL()}
 * @since 4.1.0
 * @param {Blob} blob Blob 或 File 对象
 * @returns {Promise<string>} data:URL 格式的 Base64 字符串。
 * @example
 * const aFileParts = ['<a id="a"><b id="b">hey!</b></a>']; // 一个包含DOMString的数组
 * const htmlBlob = new Blob(aFileParts, { type: 'text/html' }); // 得到 blob
 *
 * blobToDataURL(htmlBlob).then(data=>{
 *   console.log(data); // data:text/html;base64,PGEgaWQ9ImEiPjxiIGlkPSJiIj5oZXkhPC9iPjwvYT4=
 * });
 *
 * const textBlob = new Blob(aFileParts, { type: 'text/plain' });
 *
 * blobToDataURL(textBlob).then(data=>{
 *   console.log(data); // data:text/plain;base64,PGEgaWQ9ImEiPjxiIGlkPSJiIj5oZXkhPC9iPjwvYT4=
 * });
 */
function blobToDataURL(blob) {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.readAsDataURL(blob);
    // @ts-ignore
    reader.onload = () => resolve(reader.result);
    reader.onerror = (error) => reject(error);
  });
}

export default blobToDataURL;
