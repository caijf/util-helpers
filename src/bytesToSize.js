/**
 * 字节转换存储单位，保留2位小数点
 *
 * @static
 * @alias module:Processor.bytesToSize
 * @since 3.8.0
 * @param {number} bytes 字节大小
 * @returns {string} 存储单位值
 * @example
 *
 * bytesToSize(0);
 * // => 0 B
 *
 * bytesToSize(1024);
 * // => 1 KB
 *
 * bytesToSize(3.213243*1024*1024);
 * // => 3.21 MB
 *
 * bytesToSize(1024*1024*1024);
 * // => 1 GB
 */
function bytesToSize(bytes) {
  const numBytes = typeof bytes !== 'number' ? Number(bytes) : bytes;
  if (numBytes === 0 || isNaN(numBytes)) return '0 B';

  const k = 1024;
  // 存储单位
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
  const i = Math.floor(Math.log(numBytes) / Math.log(k));

  return sizes[i] ? `${Number((numBytes / Math.pow(k, i)).toFixed(2))} ${sizes[i]}` : numBytes + '';
}

export default bytesToSize;
