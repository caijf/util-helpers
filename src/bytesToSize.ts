import { isNaN } from 'ut2';

type BytesToSizeOptions = {
  spaceMark?: string;
  precision?: number;
};

/**
 * 字节转换存储单位
 *
 * @alias module:Processor.bytesToSize
 * @since 3.8.0
 * @param {number} bytes 字节大小
 * @param {Object} [options] 配置项
 * @param {string} [options.spaceMark=' '] 间隔字符，默认`' '`
 * @param {number} [options.precision=2] 精度，默认`2`
 * @returns {string} 存储单位值
 * @example
 *
 * bytesToSize(0); // 0 B
 * bytesToSize(1024); // 1 KB
 * bytesToSize(3.213243*1024*1024); // 3.21 MB
 * bytesToSize(1024*1024*1024); // 1 GB
 *
 * // 自定义间隔符号
 * bytesToSize(0, { spaceMark: '' }); // 0B
 * bytesToSize(1024, { spaceMark: '' }); // 1KB
 *
 */
function bytesToSize(bytes: number, options: BytesToSizeOptions = {}) {
  const { spaceMark = ' ', precision = 2 } = options;

  const numBytes = typeof bytes !== 'number' ? Number(bytes) : bytes;
  if (numBytes === 0 || isNaN(numBytes)) return `0${spaceMark}B`;

  const k = 1024;
  // 存储单位
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
  const i = Math.floor(Math.log(numBytes) / Math.log(k));

  return sizes[i] ? `${Number((numBytes / Math.pow(k, i)).toFixed(precision))}${spaceMark}${sizes[i]}` : numBytes + '';
}

export default bytesToSize;
