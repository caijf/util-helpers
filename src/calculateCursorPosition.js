import normalizeString from './normalizeString';

// 参考了: https://github.com/ant-design/ant-design-mobile/blob/v2/components/input-item/index.tsx#L240

/**
 * 计算输入框的值格式化后光标位置
 *
 * @static
 * @alias module:Other.calculateCursorPosition
 * @since 4.6.0
 * @see 格式化手机号码 {@link https://doly-dev.github.io/util-helpers/module-Processor.html#.formatMobile|formatMobile}
 * @see 格式化银行卡号 {@link https://doly-dev.github.io/util-helpers/module-Processor.html#.formatBankCard|formatBankCard}
 * @see h5示例 {@link https://2950v9.csb.app/|点击查看}
 * @see react示例 {@link https://33ccy9.csb.app/|点击查看}
 * @param {number} prevPos 赋值前的光标位置，onChange/onInput的光标位置 e.target.selectionEnd
 * @param {string} prevCtrlValue 上一个格式化后的值
 * @param {string} rawValue 当前输入原值
 * @param {string} ctrlValue 当前格式化后的值
 * @param {Object} [options] 配置项
 * @param {string[]|string} [options.placeholderChar=' '] 占位符
 * @param {RegExp} [options.maskReg=/\D/g] 需要遮盖的字符规则。默认去掉非数字，意味着 ctrlValue 需要去掉非数字。
 * @param {'mobile'|'bankCard'} [options.type] 格式化类型，内置手机号码和银行卡号特殊处理
 * @returns {number} 格式化后的光标位置
 */
function calculateCursorPosition(prevPos, prevCtrlValue, rawValue, ctrlValue, { placeholderChar = ' ', maskReg = /\D/g, type } = {}) {
  const realCtrlValue = normalizeString(prevCtrlValue);
  const realRawValue = normalizeString(rawValue);
  const placeholderChars = Array.isArray(placeholderChar) ? placeholderChar : [placeholderChar];

  const editLength = realRawValue.length - realCtrlValue.length;
  const isAddition = editLength > 0;

  let pos = prevPos;

  if (isAddition) {
    const additionStr = realRawValue.substring(pos - editLength, pos);
    let ctrlCharCount = additionStr.replace(maskReg, '').length;
    pos -= editLength - ctrlCharCount;

    let placeholderCharCount = 0;
    while (ctrlCharCount > 0) {
      if (placeholderChars.indexOf(ctrlValue.charAt(pos - ctrlCharCount + placeholderCharCount)) !== -1) {
        placeholderCharCount++;
      } else {
        ctrlCharCount--;
      }
    }
    pos += placeholderCharCount;
  }

  if ((type === 'mobile' && (pos === 4 || pos === 9)) || (type === 'bankCard' && pos > 0 && pos % 5 === 0)) {
    pos -= 1;
  }

  return pos;
}

export default calculateCursorPosition;
