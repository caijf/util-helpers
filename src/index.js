/**
 * 数据验证
 *
 * @module Validator
 * @since 1.1.0
 */
export { default as isMobile } from './isMobile';
export { default as isTelephone } from './isTelephone';
export { default as isPostcode } from './isPostcode';
export { default as isIdCard } from './isIdCard';
export { default as isEmail } from './isEmail';
export { default as isQQ } from './isQQ';
export { default as isWX } from './isWX';
export { default as isVehicle } from './isVehicle';
export { default as isBankCard } from './isBankCard';
export { default as isSocialCreditCode } from './isSocialCreditCode';
export { default as isPassword } from './isPassword';
export { default as isPassport } from './isPassport';
export { default as isChinese } from './isChinese';
export { default as isIPv4 } from './isIPv4';
export { default as isIPv6 } from './isIPv6';
export { default as isUrl } from './isUrl';
export { default as isBusinessLicense } from './isBusinessLicense';
export { default as validatePassword } from './validatePassword';
export { default as isPromiseLike } from './isPromiseLike';
export { default as isHMCard } from './isHMCard';
export { default as isTWCard } from './isTWCard';
export { default as isSwiftCode } from './isSwiftCode';

/**
 * 数据处理
 *
 * @module Processor
 * @since 1.1.0
 */
export { default as formatMoney } from './formatMoney';
export { default as formatBankCard } from './formatBankCard';
export { default as replaceChar } from './replaceChar';
export { default as numberToChinese } from './numberToChinese';
export { default as bytesToSize } from './bytesToSize';
export { default as parseIdCard } from './parseIdCard';
export { default as blobToDataURL } from './blobToDataURL';
export { default as dataURLToBlob } from './dataURLToBlob';
export { default as setDataURLPrefix } from './setDataURLPrefix';
export { default as normalizeString } from './normalizeString';
export { default as safeDate } from './safeDate';
export { default as formatMobile } from './formatMobile';
export { default as padZero } from './padZero';
export { default as transformFieldNames } from './transformFieldNames';
export { default as listToTree } from './listToTree';
export { default as treeToList } from './treeToList';

/**
 * 数学计算，修正浮点数计算问题
 *
 * @module Math
 * @since 3.1.0
 * @see 参考 {@link https://github.com/camsong/blog/issues/9|JavaScript 浮点数陷阱及解法}
 * @see 参考 {@link https://2zbuy.csb.app/|JS浮点数计算测试}
 * @example
 * // 从 4.12.0 版本开始，规范了有效数值。（注意：4.12.3 对有效数值重新定义）
 * // 有效数值即能通过 Number(value) 转为数字，且不能为 NaN 。
 * 
 * // 以下为有效数值：
 * // 1. 数字: 1, -1, 1e-2, 1.312, 0.1, Infinity
 * // 2. 字符串: '1', '10e2', '-1', '0.1', '', ' ', '  15', ' 15   '
 * // 4. 其他类型: null, new Date(), [], new Array(), true, false,...
 * 
 * // 以下为无效数值：
 * // 1. 字符串: '1a', '-12a', '10.2.2', '10e2.1'
 * // 2. 其他类型: undefined, [], {}, Symbol(), function(){}, ()=>{}
 * 
 * // 计算说明：
 * // 四则运算的第二个参数都是有默认值（乘数和除数默认1，加数和减数默认0）
 * plus(); // NaN  0个参数时，被加数转换为 Number(undefined) NaN ，NaN+0 = NaN 。其他计算方法如果没有参数一样返回 NaN 。
 * plus(0.1); // 0.1  第二个参数，加数默认为 0
 * plus(undefined, 0.1); // NaN  第一个参数被加数转换为 Number(undefined) NaN ，NaN+0 = NaN 。其他计算方法如果第一个参数为无效数值一样返回 NaN 。
 * plus(true, null); // 1  Number(true) 转换为 1 ， Number(null) 转换为 0 ， 1+0=1
 * 
 * // 2. 参数中包含无效数值，返回NaN
 * plus('0.1', ' a'); // NaN
 * plus(true, {}); // NaN
 * plus(true, 0.1, Symbol()); // NaN
 * 
 * // 注意：
 * // 如果第二个及后面的参数如果值为 undefined 取默认值，即乘除数取 1 ，加减法取 0 。
 * plus(0.1, undefined); // 0.1  
 * plus(0.1, undefined, 0.2, undefined); // 0.3  后面的 undefined 取默认值 0
 * times(0.1, undefined, 0.2, undefined); // 0.02  后面的 undefined 取默认值 1
 * 
 */
export { default as plus } from './plus';
export { default as minus } from './minus';
export { default as times } from './times';
export { default as divide } from './divide';
export { default as round } from './round';

/**
 * 其他
 *
 * @module Other
 * @since 4.2.0
 */
export { default as waitTime } from './waitTime';
export { default as calculateCursorPosition } from './calculateCursorPosition';
export { default as randomString } from './randomString';
export { default as strlen } from './strlen';
export { default as findTreeNode } from './findTreeNode';
export { default as findTreeSelect } from './findTreeSelect';

/**
 * 调试相关
 *
 * @module Debug
 * @ignore
 * @since 3.6.1
 */
export { setDisableWarning, version } from './utils/config';
