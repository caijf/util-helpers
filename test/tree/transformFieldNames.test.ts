import { transformFieldNames } from '../../src';

const options = [
  {
    code: '1',
    name: 'one'
  },
  {
    code: '2',
    name: 'two'
  }
];

const options2 = [
  {
    code: '1',
    name: 'one'
  },
  {
    code: '2',
    name: 'two'
  },
  {
    code: '3',
    name: 'three',
    children: [
      {
        code: '3-1',
        name: 'three-one',
        children: [
          {
            code: '3-1-1',
            name: 'three-one-one'
          },
          {
            code: '3-1-2'
            // name: 'three-one-two'
          }
        ]
      }
    ]
  }
];

const options3 = [
  {
    code: '1',
    name: 'one'
  },
  {
    code: '2',
    name: 'two'
  },
  {
    code: '3',
    name: 'three',
    // bigN: 1n,
    nan: NaN,
    func: function () {},
    arrowFunc: () => {},
    max: Infinity,
    min: -Infinity,
    // date: new Date(),
    undef: undefined,
    symbol: Symbol(),
    // [Symbol()]: 'symbol',
    // [Symbol.for('a')]: 'symbol for',
    childs: [
      {
        code: '3-1',
        name: 'three-one'
      },
      {
        code: '3-2',
        name: 'three-two',
        childs: [
          {
            code: '3-2-1',
            name: 'three-two-one'
          },
          {
            code: '3-2-2',
            name: 'three-two-two'
          }
        ]
      }
    ]
  }
];

describe('transformFieldNames', () => {
  it('return new object', () => {
    const newOpts = transformFieldNames(options, {
      label: 'name',
      value: 'code'
    });

    expect(options).toMatchObject([
      { code: '1', name: 'one' },
      { code: '2', name: 'two' }
    ]);
    expect(newOpts).toMatchObject([
      { label: 'one', value: '1' },
      { label: 'two', value: '2' }
    ]);
    expect(options).not.toEqual(newOpts);
  });

  it('incorrect', () => {
    const obj = {};
    const treeData = [{ a: 1, children: ['1'] }, 2];

    // @ts-ignore
    expect(transformFieldNames(treeData, { a: 'b' })).toMatchObject([{ a: 1, children: ['1'] }, 2]);

    // @ts-ignore
    expect(transformFieldNames(obj, {})).toBe(obj);
    expect(transformFieldNames([], {})).toMatchObject([]);
  });

  it('assign fields', () => {
    const newOpts = transformFieldNames(options, { label: 'name' });
    expect(newOpts).toMatchObject([
      { label: 'one', code: '1' },
      { label: 'two', code: '2' }
    ]);
  });

  it('recusion', () => {
    const newOpts = transformFieldNames(options2, { label: 'name', value: 'code' }, 'children');
    expect(newOpts).toMatchSnapshot();
  });

  it('complex data and custom children field name.', () => {
    const newOpts = transformFieldNames(options3, { label: 'name', children: 'childs' }, 'childs');
    expect(newOpts).toMatchSnapshot();
  });

  it('nodeAssign', () => {
    const basicMenus = [
      { id: '1', name: '首页', code: 'trade', pid: null },
      { id: '2', name: '交易管理', code: 'trade', pid: null, children: [{ id: '3', name: '交易查询', code: 'trade-1', pid: '2', children: [{ id: '4', name: '交易查询-查询操作', code: 'trade-1-1', pid: '3' }] }] },
      {
        id: '5',
        name: '权限管理',
        code: 'authorization',
        pid: null,
        children: [
          { id: '6', name: '角色管理', code: 'authorization-1', pid: '5' },
          { id: '7', name: '用户管理', code: 'authorization-2', pid: '5' }
        ]
      }
    ];

    const result = transformFieldNames(basicMenus, { label: 'name', childs: 'children' }, 'children', 'self');
    // const result = transformFieldNames(basicMenus, { label: 'name' }, undefined, 'self');

    expect(result[0]).toMatchObject(basicMenus[0]);
    expect(result[1]).toMatchObject(basicMenus[1]);
    expect(result[2]).toMatchObject(basicMenus[2]);
    expect(result).toMatchSnapshot();
  });
});
