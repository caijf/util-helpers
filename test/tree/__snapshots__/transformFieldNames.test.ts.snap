// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformFieldNames complex data and custom children field name. 1`] = `
[
  {
    "code": "1",
    "label": "one",
  },
  {
    "code": "2",
    "label": "two",
  },
  {
    "arrowFunc": [Function],
    "children": [
      {
        "code": "3-1",
        "label": "three-one",
      },
      {
        "children": [
          {
            "code": "3-2-1",
            "label": "three-two-one",
          },
          {
            "code": "3-2-2",
            "label": "three-two-two",
          },
        ],
        "code": "3-2",
        "label": "three-two",
      },
    ],
    "code": "3",
    "func": [Function],
    "label": "three",
    "max": Infinity,
    "min": -Infinity,
    "nan": NaN,
    "symbol": Symbol(),
    "undef": undefined,
  },
]
`;

exports[`transformFieldNames nodeAssign 1`] = `
[
  {
    "code": "trade",
    "id": "1",
    "label": "首页",
    "pid": null,
  },
  {
    "childs": [
      {
        "childs": [
          {
            "code": "trade-1-1",
            "id": "4",
            "label": "交易查询-查询操作",
            "pid": "3",
          },
        ],
        "code": "trade-1",
        "id": "3",
        "label": "交易查询",
        "pid": "2",
      },
    ],
    "code": "trade",
    "id": "2",
    "label": "交易管理",
    "pid": null,
  },
  {
    "childs": [
      {
        "code": "authorization-1",
        "id": "6",
        "label": "角色管理",
        "pid": "5",
      },
      {
        "code": "authorization-2",
        "id": "7",
        "label": "用户管理",
        "pid": "5",
      },
    ],
    "code": "authorization",
    "id": "5",
    "label": "权限管理",
    "pid": null,
  },
]
`;

exports[`transformFieldNames recusion 1`] = `
[
  {
    "label": "one",
    "value": "1",
  },
  {
    "label": "two",
    "value": "2",
  },
  {
    "children": [
      {
        "children": [
          {
            "label": "three-one-one",
            "value": "3-1-1",
          },
          {
            "value": "3-1-2",
          },
        ],
        "label": "three-one",
        "value": "3-1",
      },
    ],
    "label": "three",
    "value": "3",
  },
]
`;
